packages/vue/src/utils.ts

```diff
@@ -89,3 +90,109 @@ export function transformComposableProps<T extends { context: object }>(props: T
+  // HTML Comment nodes or Text nodes
+  return false
+}
+
+/**
+ * Recursively flattens the Fragment descendants of a given VNode.
+ *
+ * When you create a component and pass a <slot />, Vue wraps
+ * the contents of <slot /> inside a <Fragment /> component and assigns
+ * the <slot /> VNode a type of Fragment.
+ *
+ * So why are we flattening here? Vue renders VNodes from the leaf
+ * nodes going up to the root. In other words, when executing the render function
+ * of each component, it executes the child render functions first before the parents.
+ *
+ * This means that at any components render function execution context, all it's children
+ * VNodes should have already been rendered -- and that includes any slots! :D
+ *
+ * In the cases where we pass in a component with slots to the `asChild` component,
+ * we shall need to flatten those slot fragment VNodes so as to extract all it's children VNodes
+ * to correctly apply the props and event listeners from the with as child components.
+ *
+ * We do this recursively to ensure that all first child slots that contain fragments in their descendants are rendered into VNodes before passing events.
+ * to the first actual element VNode.
+ */
+export function renderSlotFragments(children: VNode[]): VNode[] {
+  return children.flatMap((child) => {
+    if (child.type === Fragment) {
+      return renderSlotFragments(child.children as VNode[])
+    }
+    return [child]
+  })
+}
+
+/**
+ * We use our own `mergeProps` instead of Vue's mergeProps to correctly handle event listeners.
+ *
+ * Credit: headless-ui/vue: https://github.com/tailwindlabs/headlessui/blob/main/packages/@headlessui-vue/src/utils/render.ts#L196
+ */
+export function mergeProps(...propsArgs: Record<any, any>[]) {
+  if (propsArgs.length === 0) return {}
+  if (propsArgs.length === 1) return propsArgs[0]
+
+  const target: Record<any, any> = {}
+
+  const eventHandlers: Record<
+    string,
+    ((event: { defaultPrevented: boolean }, ...args: any[]) => void | undefined)[]
+  > = {}
+
+  for (const props of propsArgs) {
+    for (const prop in props) {
+      // Collect event handlers
+      if (prop.startsWith('on') && typeof props[prop] === 'function') {
+        eventHandlers[prop] ??= []
+        eventHandlers[prop].push(props[prop])
+      } else {
+        // Override incoming prop
+        target[prop] = props[prop]
+      }
+    }
+  }
+
+  // Do not bind any event handlers when there is a `disabled` or `aria-disabled` prop set.
+  if (target.disabled || target['aria-disabled']) {
+    return Object.assign(
+      target,
+      // Set all event listeners that we collected to `undefined` so that `cloneVNode` from Vue doesn't drop them. `cloneVNode` only merges the
+      // existing and new props, they don't just override therefore we have to
+      // explicitly declare and nullify them.
+      Object.fromEntries(Object.keys(eventHandlers).map((eventName) => [eventName, undefined])),
+    )
+  }
```

Comment:
This is not an assuming we make in React or Vue. I'd rather not have us to this because it's too implicit.